### PR TITLE
[CARBONDATA-1586][Streaming] Support handoff from row format to columnar format

### PR DESCRIPTION
**Pull request by QiangCai**
1. configuration
carbon.handoff.size =>carbon.streaming.segment.max.size
2. SQL command
alter table <table_name> compact 'streaming'
3. refactory CompactionType

 - [x] Any interfaces changed?
 yes, alter table compact 'streaming'
 - [x] Any backward compatibility impacted?
 no
 - [x] Document update required?
yes
 - [x] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
         yes
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
          no
        - Any additional information to help reviewers in testing this change.
       
 - [x] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA. 
NA

### CODE REVIEW

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1862/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1437/

**Comment by QiangCai** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1440/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1441/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1863/

File: core/src/main/java/org/apache/carbondata/core/scan/result/iterator/RawResultIterator.java [hunk 1]
```diff
@@ -64,6 +64,10 @@ public RawResultIterator(CarbonIterator<BatchResult> detailRawQueryResultIterato
     this.destinationSegProperties = destinationSegProperties;
   }
 
+  public RawResultIterator() {
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/scan/result/iterator/RawResultIterator.java]: add comment for this function

File: hadoop/src/main/java/org/apache/carbondata/hadoop/streaming/CarbonStreamRecordReader.java [hunk 2]
```diff
@@ -342,7 +349,11 @@ private boolean nextRow() throws IOException {
             input.nextRow();
             scanMore = false;
           } else {
-            readRowFromStream();
+            if (useRawRow) {
+              readRawRowFromStream();
```
Comment (jackylk) [inline hadoop/src/main/java/org/apache/carbondata/hadoop/streaming/CarbonStreamRecordReader.java]: add comment to describe this is for `handoff`, which does not require decode raw row

File: integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/CarbonMergerRDD.scala [hunk 3]
```diff
@@ -116,7 +116,7 @@ class CarbonMergerRDD[K, V](
 
         // During UPDATE DELTA COMPACTION case all the blocks received in compute belongs to
         // one segment, so max cardinality will be calculated from first block of segment
-        if(carbonMergerMapping.campactionType == CompactionType.IUD_UPDDEL_DELTA_COMPACTION) {
+        if(CompactionType.IUD_UPDDEL_DELTA == carbonMergerMapping.campactionType) {
```
Comment (jackylk) [inline integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/CarbonMergerRDD.scala]: add space after `if`

File: streaming/src/main/scala/org/apache/carbondata/streaming/StreamHandoffUtil.scala [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+import org.apache.carbondata.spark.util.CommonUtil
+
+object StreamHandoffUtil {
+
+  private val LOGGER = LogServiceFactory.getLogService(this.getClass.getCanonicalName)
+
+  /**
+   * start new thread to execute stream segment handoff
+   */
+  def startStreamingHandoffThread(
```
Comment (jackylk) [inline streaming/src/main/scala/org/apache/carbondata/streaming/StreamHandoffUtil.scala]: Please move it to HandOffRDD file

File: streaming/src/main/scala/org/apache/carbondata/streaming/StreamHandoffUtil.scala [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+          if (lock.lockWithRetries()) {
+            LOGGER.info("Acquired the handoff lock for table" +
+                        s" ${ carbonTable.getDatabaseName }.${ carbonTable.getTableName }")
+            // handoff streaming segment one by one
+            do {
+              val segmentStatusManager = new SegmentStatusManager(identifier)
+              var loadMetadataDetails: Array[LoadMetadataDetails] = null
+              val statusLock = segmentStatusManager.getTableStatusLock
+              try {
+                if (statusLock.lockWithRetries()) {
```
Comment (jackylk) [inline streaming/src/main/scala/org/apache/carbondata/streaming/StreamHandoffUtil.scala]: add comment to describe what this lock is for

File: streaming/src/main/scala/org/apache/carbondata/streaming/StreamHandoffUtil.scala [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+            lock.unlock()
+          }
+        }
+      }
+    }
+    handoffThread.start()
+  }
+
+  /**
+   * invoke StreamHandoffRDD to handoff streaming segment one bye one
```
Comment (jackylk) [inline streaming/src/main/scala/org/apache/carbondata/streaming/StreamHandoffUtil.scala]: it is one only

File: integration/spark2/src/test/scala/org/apache/spark/carbondata/TestStreamingTableOperation.scala [hunk 7]
```diff
@@ -849,7 +892,13 @@ class TestStreamingTableOperation extends QueryTest with BeforeAndAfterAll {
          | salary FLOAT
          | )
          | STORED BY 'carbondata'
-         | TBLPROPERTIES(${if (streaming) "'streaming'='true', " else "" }
```
Comment (jackylk) [inline integration/spark2/src/test/scala/org/apache/spark/carbondata/TestStreamingTableOperation.scala]: I think it is not required to modify this

File: streaming/src/main/scala/org/apache/carbondata/streaming/StreamHandoffRDD.scala [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+    carbonLoadModel: CarbonLoadModel,
+    handOffSegmentId: String
+) extends CarbonRDD[(K, V)](sc, Nil) {
+
+  private val jobTrackerId: String = {
+    val formatter = new SimpleDateFormat("yyyyMMddHHmm")
+    formatter.format(new Date())
+  }
+
+  override def internalCompute(split: Partition,
```
Comment (jackylk) [inline streaming/src/main/scala/org/apache/carbondata/streaming/StreamHandoffRDD.scala]: move `split` to next line

File: streaming/src/main/scala/org/apache/carbondata/streaming/StreamHandoffRDD.scala [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+  }
+
+  private def prepareHandoffProcessor(
+      carbonTable: CarbonTable
+  ): CompactionResultSortProcessor = {
+    val wrapperColumnSchemaList = CarbonUtil.getColumnSchemaList(
+      carbonTable.getDimensionByTableName(carbonTable.getTableName),
+      carbonTable.getMeasureByTableName(carbonTable.getTableName))
+    val dimLensWithComplex = new Array[Int](wrapperColumnSchemaList.size())
+    for (i <- 0 until dimLensWithComplex.length) {
```
Comment (jackylk) [inline streaming/src/main/scala/org/apache/carbondata/streaming/StreamHandoffRDD.scala]: use dimLensWithComplex.length.zipWithIndex

File: streaming/src/main/scala/org/apache/carbondata/streaming/StreamHandoffRDD.scala [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+    val formatter = new SimpleDateFormat("yyyyMMddHHmm")
+    formatter.format(new Date())
+  }
+
+  override def internalCompute(split: Partition,
+      context: TaskContext): Iterator[(K, V)] = {
+    carbonLoadModel.setPartitionId("0")
+    carbonLoadModel.setTaskNo("" + split.index)
+    val carbonTable = carbonLoadModel.getCarbonDataLoadSchema.getCarbonTable
+    val iteratorList = prepareInputIterator(split, carbonTable)
```
Comment (jackylk) [inline streaming/src/main/scala/org/apache/carbondata/streaming/StreamHandoffRDD.scala]: add comment to describe it is using raw row

File: processing/src/main/java/org/apache/carbondata/processing/merger/CompactionResultSortProcessor.java [hunk 11]
```diff
@@ -196,9 +196,16 @@ private void deleteTempStoreLocation() {
    */
   private void processResult(List<RawResultIterator> resultIteratorList) throws Exception {
     for (RawResultIterator resultIterator : resultIteratorList) {
-      while (resultIterator.hasNext()) {
-        addRowForSorting(prepareRowObjectForSorting(resultIterator.next()));
-        isRecordFound = true;
+      if (CompactionType.STREAMING == compactionType) {
+        while (resultIterator.hasNext()) {
+          addRowForSorting(resultIterator.next());
```
Comment (jackylk) [inline processing/src/main/java/org/apache/carbondata/processing/merger/CompactionResultSortProcessor.java]: add comment to describe it is using raw row

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1474/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1893/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1480/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1481/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1899/

**Pull request closed without merging**